src/async.zsh

```diff
@@ -9,49 +9,59 @@ _zsh_autosuggest_async_request() {
 	typeset -g _ZSH_AUTOSUGGEST_ASYNC_FD _ZSH_AUTOSUGGEST_CHILD_PID
 
 	# If we've got a pending request, cancel it
-	if [[ -n "$_ZSH_AUTOSUGGEST_ASYNC_FD" ]] && { true <&$_ZSH_AUTOSUGGEST_ASYNC_FD } 2>/dev/null; then
-		# Close the file descriptor and remove the handler
-		exec {_ZSH_AUTOSUGGEST_ASYNC_FD}<&-
-		zle -F $_ZSH_AUTOSUGGEST_ASYNC_FD
-
-		# We won't know the pid unless the user has zsh/system module installed
-		if [[ -n "$_ZSH_AUTOSUGGEST_CHILD_PID" ]]; then
-			# Zsh will make a new process group for the child process only if job
-			# control is enabled (MONITOR option)
-			if [[ -o MONITOR ]]; then
-				# Send the signal to the process group to kill any processes that may
-				# have been forked by the suggestion strategy
-				kill -TERM -$_ZSH_AUTOSUGGEST_CHILD_PID 2>/dev/null
+	if (( _ZSH_AUTOSUGGEST_CHILD_PID )); then
+		kill -TERM -- $_ZSH_AUTOSUGGEST_CHILD_PID 2>/dev/null
+		_ZSH_AUTOSUGGEST_CHILD_PID=
+	fi
+
+	_ZSH_AUTOSUGGEST_ASYNC_FD=
+
+	{
+		# Fork a process to fetch a suggestion and open a pipe to read from it
+		exec {_ZSH_AUTOSUGGEST_ASYNC_FD}< <(
+			# Suppress error messages
+			exec 2>/dev/null
+
+			# Tell parent process our pid
+			if (( ${+sysparams} )); then
+				echo ${sysparams[pid]} || return
 			else
-				# Kill just the child process since it wasn't placed in a new process
-				# group. If the suggestion strategy forked any child processes they may
-				# be orphaned and left behind.
-				kill -TERM $_ZSH_AUTOSUGGEST_CHILD_PID 2>/dev/null
+				echo || return
 			fi
-		fi
-	fi
 
-	# Fork a process to fetch a suggestion and open a pipe to read from it
-	exec {_ZSH_AUTOSUGGEST_ASYNC_FD}< <(
-		# Tell parent process our pid
-		echo $sysparams[pid]
+			# Fetch and print the suggestion
+			local suggestion
+			_zsh_autosuggest_fetch_suggestion "$1"
+			echo -nE - "$suggestion"
+		) || return
```

Comment:
@romkatv Can you elaborate on the addition of the ` || return` added here and after `read` and `echo`? In what situations can `exec`/`read`/`echo` have non-zero exit status?

Reply:
I suppose one could read the source code of the builtins `exec`/`read`/`echo` and find out which syscalls they rely on. Then read the source code or docs of all operating systems to figure out when those system calls can fail.

I think a better question is *what* the code should do when `exec`/`read`/`echo` fail. Which variant of this code do you think behaves better in case of errors?